tools/p4mlir-translate/translate.cpp

```diff
@@ -467,6 +558,20 @@ class P4HIRConverter : public P4::Inspector, public P4::ResolutionContext {
     mlir::Value emitBinOp(const P4::IR::Operation_Binary *binop, P4HIR::BinOpKind kind);
     mlir::Value emitConcatOp(const P4::IR::Concat *concatop);
     mlir::Value emitCmp(const P4::IR::Operation_Relation *relop, P4HIR::CmpOpKind kind);
+
+ private:
+    mlir::Value emitInvalidHeaderCmpOp(const P4::IR::Operation_Relation *p4RelationOp);
+    mlir::Value emitInvalidHeaderUnionCmpOp(const P4::IR::Operation_Relation *p4RelationOp);
+    mlir::Value emitHeaderBuiltInMethod(mlir::Location loc, const P4::BuiltInMethod *builtInMethod);
+    mlir::Value emitHeaderUnionBuiltInMethod(mlir::Location loc,
+                                             const P4::BuiltInMethod *builtInMethod);
+    template <typename DialectType>
+    DialectType getType(mlir::Value &value) {
+        return mlir::isa<P4HIR::ReferenceType>(value.getType())
```

Comment:
```suggestion
        if (auto refType = mlir::dyn_cast<P4HIR::ReferenceType>(value.getType()))
           return refType.getObjectType();
        return value.getType();
```

The explicit cast is quite misleading here IMO. I would just make `getObjectType` return `mlir::Type` and perform cast at caller side.

Reply:
Done